src/index.js

```diff
@@ -101,6 +101,7 @@ export { default as reduceP } from './reduceP';
 export { default as reduceRightP } from './reduceRightP';
 export { default as sliceFrom } from './sliceFrom';
 export { default as sliceTo } from './sliceTo';
+export { default as updateSlice } from './updateSlice';
```

Comment:
Typings:
```ts
updateSlice(fromIndex: number, toIndex: number, replacement: string|any[], list: string|any[]): string|any[];
updateSlice(fromIndex: number, toIndex: number, replacement: string|any[]) =>  list: string|any[]): string|any[];
updateSlice(fromIndex: number, toIndex: number): {
  (replacement: string|any[], list: string|any[]) => string|any[];
  (replacement: string|any[]):  (list: string|any[]) => string|any[];
}
updateSlice(fromIndex: number): {
  (toIndex: number, replacement: string|any[], list: string|any[]) => string|any[];
  (toIndex: number, replacement: string|any[]): (list: string|any[]) => string|any[];
  (toIndex: number): {
    (replacement: string|any[], list: string|any[]) => string|any[];
    (replacement: string|any[]): (list: string|any[]) => string|any[];
  }
}
```

@BjornMelgaard can you double check pls ?

Reply:
IMO, it's better using extends
```ts
updateSlice<T extends string|any[]>(fromIndex: number, toIndex: number, replacement: T, list: T): T;
updateSlice<T extends string|any[]>(fromIndex: number, toIndex: number, replacement: T) =>  list: T): T;
updateSlice<T extends string|any[]>(fromIndex: number, toIndex: number): {
  (replacement: T, list: T) => T;
  (replacement: T):  (list: T) => T;
}
updateSlice<T extends string|any[]>(fromIndex: number): {
  (toIndex: number, replacement: T, list: T) => T;
  (toIndex: number, replacement: T): (list: T) => T;
  (toIndex: number): {
    (replacement: T, list: T) => T;
    (replacement: T): (list: T) => T;
  }
}
```

Reply:
@guillaumear

> When I tried to do something like (simplified): `<S extends string|any[]>(replacement: S, target: S): S`, when I passed string literals "a" and "s", it inferred S to be "a" | "s" ._.

Reply:
better not use extend [because you cant use `updateSlice(1, 3, 'DDDD', 'ABBCCC') == 'adsf'
`](https://www.typescriptlang.org/play/index.html#src=declare%20function%20updateSlice%3C%20T%20extends%20string%20%7C%20any%5B%5D%20%3E%20(fromIndex%3A%20number%2C%20toIndex%3A%20number%2C%20replacement%3A%20T%2C%20list%3A%20T)%3A%20T%3B%0D%0Adeclare%20function%20updateSlice%3CT%20extends%20string%7Cany%5B%5D%3E(fromIndex%3A%20number%2C%20toIndex%3A%20number%2C%20replacement%3A%20T)%3A%20%20(list%3A%20T)%20%3D%3E%20T%3B%0D%0Adeclare%20function%20updateSlice%3CT%20extends%20string%7Cany%5B%5D%3E(fromIndex%3A%20number%2C%20toIndex%3A%20number)%3A%20%7B%0D%0A%20%20(replacement%3A%20T%2C%20list%3A%20T)%3A%20T%3B%0D%0A%20%20(replacement%3A%20T)%3A%20%20(list%3A%20T)%20%3D%3E%20T%3B%0D%0A%7D%0D%0Adeclare%20function%20updateSlice%3CT%20extends%20string%7Cany%5B%5D%3E(fromIndex%3A%20number)%3A%20%7B%0D%0A%20%20(toIndex%3A%20number%2C%20replacement%3A%20T%2C%20list%3A%20T)%3A%20T%3B%0D%0A%20%20(toIndex%3A%20number%2C%20replacement%3A%20T)%3A%20(list%3A%20T)%20%3D%3E%20T%3B%0D%0A%20%20(toIndex%3A%20number)%3A%20%7B%0D%0A%20%20%20%20(replacement%3A%20T%2C%20list%3A%20T)%3A%20T%3B%0D%0A%20%20%20%20(replacement%3A%20T)%3A%20(list%3A%20T)%20%3D%3E%20T%3B%0D%0A%20%20%7D%0D%0A%7D%0D%0A%0D%0AupdateSlice(1%2C%203%2C%20'DDDD'%2C%20'ABBCCC')%20%2F%2F%3D%3E%20'ADDDDCCC'%0D%0A%0D%0A%0D%0AupdateSlice(1%2C%203%2C%20'DDDD')('ABBCCC')%20%2F%2F%3D%3E%20'ADDDDCCC'%0D%0AupdateSlice(1%2C%203)('DDDD'%2C%20'ABBCCC')%20%2F%2F%3D%3E%20'ADDDDCCC'%0D%0AupdateSlice(1)(3%2C%20'DDDD'%2C%20'ABBCCC')%20%2F%2F%3D%3E%20'ADDDDCCC'%0D%0A%0D%0A%0D%0AupdateSlice(1)(3)('DDDD'%2C%20'ABBCCC')%20%2F%2F%3D%3E%20'ADDDDCCC'%0D%0AupdateSlice(1)(3)('DDDD'%2C%20'ABBCCC')%20%2F%2F%3D%3E%20'ADDDDCCC'%0D%0AupdateSlice(1%2C%203)('DDDD')('ABBCCC')%20%2F%2F%3D%3E%20'ADDDDCCC'%0D%0A%0D%0A%0D%0AupdateSlice(1)(3)('DDDD')('ABBCCC')%20%2F%2F%3D%3E%20'ADDDDCCC'%0D%0A%0D%0A%0D%0A%2F%2F%20dont%20use%20extend%20because%0D%0AupdateSlice(1%2C%203%2C%20'DDDD'%2C%20'ABBCCC')%20%3D%3D%20'adsf'%0D%0A)

[this works](https://www.typescriptlang.org/play/index.html#src=declare%20function%20updateSlice(fromIndex%3A%20number%2C%20toIndex%3A%20number%2C%20replacement%3A%20string%7Cany%5B%5D%2C%20list%3A%20string%7Cany%5B%5D)%3A%20string%7Cany%5B%5D%3B%0D%0Adeclare%20function%20updateSlice(fromIndex%3A%20number%2C%20toIndex%3A%20number%2C%20replacement%3A%20string%7Cany%5B%5D)%3A%20%20(list%3A%20string%7Cany%5B%5D)%20%3D%3E%20string%7Cany%5B%5D%3B%0D%0Adeclare%20function%20updateSlice(fromIndex%3A%20number%2C%20toIndex%3A%20number)%3A%20%7B%0D%0A%20%20(replacement%3A%20string%7Cany%5B%5D%2C%20list%3A%20string%7Cany%5B%5D)%3A%20string%7Cany%5B%5D%3B%0D%0A%20%20(replacement%3A%20string%7Cany%5B%5D)%3A%20%20(list%3A%20string%7Cany%5B%5D)%20%3D%3E%20string%7Cany%5B%5D%3B%0D%0A%7D%0D%0Adeclare%20function%20updateSlice(fromIndex%3A%20number)%3A%20%7B%0D%0A%20%20(toIndex%3A%20number%2C%20replacement%3A%20string%7Cany%5B%5D%2C%20list%3A%20string%7Cany%5B%5D)%3A%20string%7Cany%5B%5D%3B%0D%0A%20%20(toIndex%3A%20number%2C%20replacement%3A%20string%7Cany%5B%5D)%3A%20(list%3A%20string%7Cany%5B%5D)%20%3D%3E%20string%7Cany%5B%5D%3B%0D%0A%20%20(toIndex%3A%20number)%3A%20%7B%0D%0A%20%20%20%20(replacement%3A%20string%7Cany%5B%5D%2C%20list%3A%20string%7Cany%5B%5D)%3A%20string%7Cany%5B%5D%3B%0D%0A%20%20%20%20(replacement%3A%20string%7Cany%5B%5D)%3A%20(list%3A%20string%7Cany%5B%5D)%20%3D%3E%20string%7Cany%5B%5D%3B%0D%0A%20%20%7D%0D%0A%7D%0D%0A%0D%0AupdateSlice(1%2C%203%2C%20'DDDD'%2C%20'ABBCCC')%20%2F%2F%3D%3E%20'ADDDDCCC'%0D%0A%0D%0A%0D%0AupdateSlice(1%2C%203%2C%20'DDDD')('ABBCCC')%20%2F%2F%3D%3E%20'ADDDDCCC'%0D%0AupdateSlice(1%2C%203)('DDDD'%2C%20'ABBCCC')%20%2F%2F%3D%3E%20'ADDDDCCC'%0D%0AupdateSlice(1)(3%2C%20'DDDD'%2C%20'ABBCCC')%20%2F%2F%3D%3E%20'ADDDDCCC'%0D%0A%0D%0A%0D%0AupdateSlice(1)(3)('DDDD'%2C%20'ABBCCC')%20%2F%2F%3D%3E%20'ADDDDCCC'%0D%0AupdateSlice(1)(3)('DDDD'%2C%20'ABBCCC')%20%2F%2F%3D%3E%20'ADDDDCCC'%0D%0AupdateSlice(1%2C%203)('DDDD')('ABBCCC')%20%2F%2F%3D%3E%20'ADDDDCCC'%0D%0A%0D%0A%0D%0AupdateSlice(1)(3)('DDDD')('ABBCCC')%20%2F%2F%3D%3E%20'ADDDDCCC'%0D%0A)